apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletVersionInstrumentation.java

```diff
@@ -0,0 +1,164 @@
+
+    @VisibleForAdvice
+    public static AtomicBoolean doLogAndWarn = new AtomicBoolean(true);
+
+    @Override
+    public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
+        return nameContains("Servlet").or(nameContainsIgnoreCase("jsp"));
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return not(isInterface())
+            .and(hasSuperType(named("javax.servlet.Servlet")));
+    }
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Collections.singleton(SERVLET_API);
+    }
+
+    /**
+     * Instruments {@link javax.servlet.Servlet#init(ServletConfig)}
+     */
+    public static class Init extends ServletVersionInstrumentation {
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+            return named("init")
+                .and(takesArgument(0, named("javax.servlet.ServletConfig")));
+        }
+
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        @SuppressWarnings("Duplicates") // duplication is fine here as it allows to inline code
+        private static void onEnter(@Advice.Argument(0) @Nullable ServletConfig servletConfig) {
+            if (!doLogAndWarn.getAndSet(false)) {
+                return;
+            }
+
+            int majorVersion = -1;
+            int minorVersion = -1;
+            String serverInfo = null;
+            if (servletConfig != null) {
+                ServletContext servletContext = servletConfig.getServletContext();
+                if (null != servletContext) {
+                    majorVersion = servletContext.getMajorVersion();
+                    minorVersion = servletContext.getMinorVersion();
+                    serverInfo = servletContext.getServerInfo();
+                }
+            }
+
+            logger.info("Servlet container info = {}", serverInfo);
+            if (majorVersion < 3) {
+                logger.warn("Unsupported servlet version detected: {}.{}, no Servlet transaction will be created", majorVersion, minorVersion);
+            }
+        }
+    }
+
+    /**
+     * Instruments {@link javax.servlet.Servlet#service(ServletRequest, ServletResponse)}
+     */
+    public static class Service extends ServletVersionInstrumentation {
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+            return named("service")
+                .and(takesArgument(0, named("javax.servlet.ServletRequest")))
+                .and(takesArgument(1, named("javax.servlet.ServletResponse")));
+        }
+
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        @SuppressWarnings("Duplicates") // duplication is fine here as it allows to inline code
+        private static void onEnter(@Advice.This Servlet servlet) {
+            if (!doLogAndWarn.getAndSet(false)) {
```

Comment:
To avoid a CAS on every invocation
```suggestion
            if (!doLogAndWarn.get() || !doLogAndWarn.getAndSet(false)) {
```

maybe rename `doLogAndWarn` to `alreadyLogged` to get rid of the negations which are a bit harder to parse?

Reply:
definitely agree on having less negations, however doing a double check here seems a bit overkill. I think it's fine to replace this `AtomicBoolean` with a simple volatile boolean as `AtomicBoolean.get` does makes a volatile read, thus it won't be more efficient.